docs/edgeql/insert.rst

```diff
@@ -0,0 +1,335 @@
+
+In the previous query, we selected Black Widow twice: once in the
+``characters`` set and again as the ``nemesis`` of Dreykov. In circumstances
+like this, you should pull that subquery into a ``with`` block.
+
+.. code-block:: edgeql-repl
+
+  db> with black_widow := (select Hero filter .name = "Black Widow")
+  ... insert Movie {
+  ...   title := "Black Widow",
+  ...   characters := {
+  ...     black_widow,
+  ...     (insert Hero { name := "Yelena Belova"}),
+  ...     (insert Villain {
+  ...       name := "Dreykov",
+  ...       nemesis := black_widow
+  ...     })
+  ...   }
+  ... }
+  {default::Movie {id: af706c7c-3e98-11ec-abb3-4bbf3f18a61a}}
+
+
+The ``with`` block can contain an arbitrary number of clauses; later clauses
+can reference earlier ones.
+
+.. code-block:: edgeql-repl
+
+  db> with
+  ...  black_widow := (select Hero filter .name = "Black Widow"),
+  ...  yelena := (insert Hero { name := "Yelena Belova"}),
+  ...  dreykov := (insert Villain {name := "Dreykov", nemesis := black_widow})
+  ... insert Movie {
+  ...   title := "Black Widow",
+  ...   characters := { black_widow, yelena, dreykov }
+  ... }
+  {default::Movie {id: af706c7c-3e98-11ec-abb3-4bbf3f18a61a}}
+
+
+.. _ref_eql_insert_conflicts:
+
+Conflicts
+---------
+
+EdgeDB provides a general-purpose mechanism for gracefully handling possible
+exclusivity constraint violations. Consider a scenario where we are trying to
+``insert`` Eternals (the ``Movie``), but we can't remember if it already exists
+in the database.
+
+.. code-block:: edgeql-repl
+
+  db> with eternals := (
+  ...   insert Movie {
+  ...     title := "Eternals"
+  ...   }
+  ...   unless conflict on .title
+  ...   else (select Movie filter .title = "Eternals")
+  ... )
+  ... insert Movie {
+  ...   title := "Black Widow",
+  ...   characters := { black_widow }
+  ... }
+  {default::Movie {id: af706c7c-3e98-11ec-abb3-4bbf3f18a61a}}
+
+In essence, ``unless conflict`` can provide a fallback expression in case an
+``insert`` fails. This query attempts to ``insert`` Eternals. If it already
+exists in the database, it will violate the uniqueness constraint on
+``Movie.title``, causing a *conflict* on the ``title`` field. The ``else``
+clause is then executed and returned instead.
+
+
+.. _ref_eql_upsert:
+
+Upserts
+^^^^^^^
+
+There are no limitations on what the ``else`` clause can contain; it can be any
+EdgeQL expression, including an :ref:`update <ref_eql_update>` statement. This
+lets you express *upsert* logic in a single EdgeQL query.
+
+.. code-block:: edgeql-repl
+
+  db> with
+  ...   title := "Eternals",
+  ...   release_year := 2021
+  ... insert Movie {
+  ...   title := title,
+  ...   release_year := release_year
+  ... }
+  ... unless conflict on .title
+  ... else (
+  ...   update Movie set { release_year := release_year }
+  ... );
+  {default::Movie {id: f1bf5ac0-3e9d-11ec-b78d-c7dfb363362c}}
+
+When a conflict occurs during the initial ``insert``, the statement falls back
```

Comment:
I'd clarify here that the `update` statement does not require a `filter`, because the `Movie` reference will always be bound to the conflicting object.